firewood/src/merkle/changes.rs

```diff
@@ -0,0 +1,389 @@
+                // If this node is a leaf, then we don't need to save its state to `node_state` as
+                // it has no children to traverse.
+                if let Node::Branch(_branch) = &*state.node {
+                    // Check if this node's path is a prefix of the key. If it is not (`unique_node`
+                    // is not empty), then this node's children cannot be larger than or equal to
+                    // the key, and we don't need to include them on the traversal stack.
+                    let path_overlap = PrefixOverlap::from(key, node_key.as_ref());
+                    let unique_node = path_overlap.unique_b;
+                    if unique_node.is_empty() {
+                        self.node_state = Some(state);
+                    }
+                }
+            } else {
+                // Traversal stack is empty. This means the key is lexicographically larger than
+                // all of the keys in the trie. Calling `next` will return None.
+                return Ok(());
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+#[allow(
+    clippy::unwrap_used,
+    clippy::arithmetic_side_effects,
+    clippy::type_complexity
+)]
+mod tests {
+    use crate::{
+        db::BatchOp,
+        iter::key_from_nibble_iter,
+        merkle::{Key, Merkle, changes::PreOrderIterator},
+    };
+
+    use firewood_storage::{
+        HashedNodeReader, ImmutableProposal, MemStore, MutableProposal, NodeStore, SeededRng,
+    };
+    use std::{collections::HashSet, sync::Arc};
+
+    fn create_test_merkle() -> Merkle<NodeStore<MutableProposal, MemStore>> {
+        let memstore = MemStore::new(vec![]);
+        let nodestore = NodeStore::new_empty_proposal(Arc::new(memstore));
+        Merkle::from(nodestore)
+    }
+
+    fn make_immutable(
+        merkle: Merkle<NodeStore<MutableProposal, MemStore>>,
+    ) -> Merkle<NodeStore<Arc<ImmutableProposal>, MemStore>> {
+        merkle.try_into().unwrap()
+    }
+
+    fn random_key_from_hashset<'a>(
+        rng: &'a SeededRng,
+        set: &'a HashSet<Vec<u8>>,
+    ) -> Option<Vec<u8>> {
+        let len = set.len();
+        if len == 0 {
+            return None;
+        }
+        // Generate a random index within the bounds of the set's size
+        let index = rng.random_range(0..len);
+        // Use the iterator's nth method to get the element at that index
+        set.iter().nth(index).cloned()
+    }
+
+    fn gen_delete_key(
+        rng: &SeededRng,
+        committed_keys: &HashSet<Vec<u8>>,
+        seen_keys: &mut HashSet<Vec<u8>>,
+    ) -> Option<Vec<u8>> {
+        let key_opt = random_key_from_hashset(rng, committed_keys);
+        // Just skip the key if it has been seen. Otherwise return it and add the key
+        // to seen keys.
+        key_opt
+            .filter(|del_key| !seen_keys.contains(del_key))
+            .inspect(|del_key| {
+                seen_keys.insert(del_key.clone());
+            })
+    }
+
+    fn gen_random_keys(
```

Comment:
name here is misleading. This doesn't just generate random keys. It generates random(ish) BatchOps.
```suggestion
    fn gen_random_test_batchops(
```